R/zchunk_L100.Population_downscale_ctry.R

```diff
@@ -34,19 +34,129 @@ module_socioeconomics_L100.Population_downscale_ctry_DISABLED <- function(comman
+      complete(iso, nesting(year)) %>%  # Completes tibble to include all years for all iso
+      group_by(iso) %>%
+      mutate(pop = if_else((year %in% socioeconomics.MADDISON_HISTORICAL_YEARS), (pop[year == socioeconomics.AGG_BASE_YEAR] * ratio_iso), pop)) %>%  # Multiply pop in 1950 by ratio to get prior years' pop
+      filter(iso != "scg") %>%  # "Serbia & Montenegro" separated into individual countries, do not need combined country
+      mutate(pop = if_else(is.na(pop), 0, pop)) %>%  # Per old data system: At this point the remaining mismatched countries are small, so set to zero
+      mutate(year = as.integer(year)) %>%  # To avoid problems later on, save years as integers
+      rename(value = pop) %>%
+      select(-ratio_iso)
+
+    ## (2) SSP population projections by country
+
+    # Final historical population from UN
+    pop_final_hist <- filter(L100.Pop_thous_ctry_Yh, year == socioeconomics.FINAL_HIST_YEAR) %>%
+      rename(pop_final_hist = value) %>%
+      select(-year)
+
+    # First, generate ratios of future population to base year (2010) for all SSPs. The ratios will be applied to the historical year populations so there are no jumps/inconsistencies.
+    L100.Pop_thous_SSP_ctry_Yfut <- as_tibble(SSP_database_v9) %>% # Note units in SSP database are millions, but convert to thousands when we multiply by historic year
+      filter(MODEL == "IIASA-WiC POP" & VARIABLE == "Population") %>%  # IIASA-WiC is the official SSP population data set
+      mutate(iso = tolower(REGION), scenario = substr(SCENARIO, 1,4)) %>%
+      select(-MODEL, -VARIABLE, -UNIT, -REGION, -SCENARIO) %>%
+      mutate(iso = gsub("rou", "rom", iso)) %>%  # SSP uses "rou" for the iso for Romania; replace with "rom" for consistency with other data sources
+      gather(year, pop, -iso, -scenario) %>%  # Long format
+      mutate(year = as.integer(year), pop = as.numeric(pop)) %>%  # Clean year variable
+      filter(year %in% c(socioeconomics.FINAL_HIST_YEAR, modeltime.FUTURE_YEARS)) %>% # Retain only years needed for GCAM
+      group_by(scenario, iso) %>%
+      mutate(ratio_iso_ssp = pop / pop[year == socioeconomics.FINAL_HIST_YEAR]) %>%  # Calculate population ratios to final historical year (2010), no units
+      select(-pop) %>%
+      # Second, project country population values using SSP ratios and final historical year populations.
+      # Not all countries in the UN data are in SSP data. Create complete tibble with all UN countries & SSP years.
+      ungroup() %>%
+      complete(scenario = unique(scenario),
+               year = unique(year),
+               iso = unique(L100.Pop_thous_ctry_Yh$iso)) %>%
+      # For these countries, the ratio will be set to 1 (per the old data system).
+      mutate(ratio_iso_ssp = if_else(is.na(ratio_iso_ssp), 1, ratio_iso_ssp)) %>%
+      ## Note: In the old data system, Taiwan is in this category and has constant population. Issue has been opened to deal with this later. ##
```

Comment:
note that i also separated the pure clean up of original data vs. ratio processing.
this way, when we get the next-gen SSPs, we only need to change the data clean up part. ratio processing need not change.